components/places/src/bookmark_sync/store.rs

```diff
@@ -0,0 +1,1020 @@
+        Ok(())
+    }
+
+    fn store_folder(
+        &self,
+        staging: Staging,
+        modified: ServerTimestamp,
+        b: FolderRecord,
+    ) -> Result<()> {
+        let needs_merge = staging == Staging::Incoming;
+        self.db.execute_named_cached(
+            r#"REPLACE INTO moz_bookmarks_synced(guid, parentGuid, serverModified, needsMerge, kind,
+                                                 dateAdded, title, validity)
+               VALUES(:guid, :parentGuid, :serverModified, :needsMerge, :kind,
+                      :dateAdded, NULLIF(:title, ""), :validity)"#,
+            &[
+                (":guid", &b.guid.as_ref()),
+                (":parentGuid", &b.parent_guid.as_ref()),
+                (":serverModified", &(modified.as_millis() as i64)),
+                (":needsMerge", &needs_merge),
+                (":kind", &SyncedBookmarkKind::Folder),
+                (":dateAdded", &b.date_added),
+                (":title", &maybe_truncate_title(&b.title)),
+                (":validity", &SyncedBookmarkValidity::Valid),
+            ],
+        )?;
+        for (position, child_guid) in b.children.iter().enumerate() {
+            self.db.execute_named_cached(
+                "REPLACE INTO moz_bookmarks_synced_structure(guid, parentGuid, position)
+                 VALUES(:guid, :parentGuid, :position)",
+                &[
+                    (":guid", &child_guid),
+                    (":parentGuid", &b.guid.as_ref()),
+                    (":position", &(position as i64)),
+                ],
+            )?;
+        }
+        Ok(())
+    }
+
+    fn maybe_store_url(&self, url: Option<&String>) -> Result<Url> {
+        if let Some(url) = url {
+            let url = Url::parse(url)?;
+            if url.as_str().len() > URL_LENGTH_MAX {
+                return Err(ErrorKind::InvalidPlaceInfo(InvalidPlaceInfo::UrlTooLong).into());
+            }
+            self.db.execute_named_cached(
+                "INSERT OR IGNORE INTO moz_places(guid, url, url_hash)
+                 VALUES(IFNULL((SELECT guid FROM moz_places
+                                WHERE url_hash = hash(:url) AND
+                                      url = :url),
+                        generate_guid()), :url, hash(:url))",
+                &[(":url", &url.as_str())],
+            )?;
+            Ok(url)
+        } else {
+            Err(ErrorKind::InvalidPlaceInfo(InvalidPlaceInfo::NoUrl).into())
+        }
+    }
+
+    pub fn apply_payload(
+        &self,
+        timestamp: ServerTimestamp,
+        payload: sync15::Payload,
+    ) -> Result<()> {
+        let item = BookmarkItemRecord::from_payload(payload)?;
+        match item {
+            BookmarkItemRecord::Bookmark(b) => {
+                self.store_bookmark(Staging::Incoming, timestamp, b)?
+            }
+            BookmarkItemRecord::Folder(f) => self.store_folder(Staging::Incoming, timestamp, f)?,
+            _ => unimplemented!("TODO: Store other types"),
```

Comment:
Let's do this next...

Reply:
This is now largely done, not withstanding a more detailed review of the dumb stuff I added :)